token/src/token.rs

```diff
@@ -0,0 +1,89 @@
+use bulletproofs::{BulletproofGens, PedersenGens};
+use curve25519_dalek::scalar::Scalar;
+use zkvm::*;
+
+pub struct Token {
+    flavor: Predicate,
+    metadata: Vec<u8>,
+    nonce: Scalar,
```

Comment:
`nonce` is a per-transaction thing, and is even optional. Ideally the Token API does not even have to worry about it unless the token's issuance predicate tries to build a singleton or issue-once logic. It's the "transaction builder"'s business to add `nonce` if necessary to a final tx — if there are some `input`s added, `nonce` is not necessary.